Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h

```diff
@@ -0,0 +1,603 @@
+  Path compute_shortest_noncontractible_cycle(typename WeightFunctor::Weight_t* length,
+                                              bool display_time=false)
+  { return compute_shortest_noncontractible_cycle(length, WeightFunctor(), display_time); }
+
+  Path compute_edgewidth(bool display_time=false)
+  { return compute_shortest_noncontractible_cycle(nullptr, display_time); }
+
+protected:
+  int vertex_info(Dart_handle dh) const
+  { return get_local_map().template info<0>(dh); }
+  int& vertex_info(Dart_handle dh)
+  { return get_local_map().template info<0>(dh); }
+
+  void create_vertex_info()
+  {
+    for (auto it=get_local_map().darts().begin(),
+           itend=get_local_map().darts().end(); it!=itend; ++it)
+    {
+      if (get_local_map().template attribute<0>(it)==nullptr)
+      { get_local_map().template set_attribute<0>
+          (it, get_local_map().template create_attribute<0>(-1)); }
+    }
+  }
+
+  void initialize_vertex_info()
+  {
+    for (auto it=get_local_map().template attributes<0>().begin(),
+           itend = get_local_map().template attributes<0>().end(); it != itend; ++it)
+    { get_local_map().template info_of_attribute<0>(it)=-1; }
+  }
+
+  template <class WeightFunctor, class Distance_>
+  void compute_spanning_tree(Dart_handle root, Dart_container& spanning_tree,
+                             std::vector<Distance_>& distance_from_root,
+                             std::vector<int>& trace_index,
+                             const WeightFunctor& wf)
+  {
+    if (std::is_same<WeightFunctor, Unit_weight_functor>::value)
+    { compute_BFS_tree(root, spanning_tree, distance_from_root, trace_index, wf); }
+    else
+    { compute_Dijkstra_tree(root, spanning_tree, distance_from_root, trace_index, wf); }
+  }
+
+  template <class Distance_>
+  struct Dijkstra_comparator
+  {
+    using Distance_type=Distance_;
+    Dijkstra_comparator(const std::vector<Distance_type>& distance_from_root) :
+      m_distance(distance_from_root) {}
+    bool operator()(const int x, const int y) const
+    { return m_distance[x]>m_distance[y]; }
+  private:
+    const std::vector<Distance_type>& m_distance;
+  };
+
+  /// Create a spanning tree using Dijkstra
+  template <class WeightFunctor, class Distance_>
+  void compute_Dijkstra_tree(Dart_handle root, Dart_container& spanning_tree,
+                             std::vector<Distance_>& distance_from_root,
+                             std::vector<int>& trace_index,
+                             const WeightFunctor& wf)
+  {
+    // Preparation
+    Dijkstra_comparator<Distance_> dc(distance_from_root);
+    std::priority_queue<int, std::vector<int>, Dijkstra_comparator<Distance_> > pq(dc);
+    int vertex_index=0;
+    size_type vertex_visited;
+    try
+    {
+      vertex_visited=get_local_map().get_new_mark();
+    }
+    catch (typename Local_map::Exception_no_more_available_mark)
+    {
+      std::cerr<<"No more free mark, exit."<<std::endl;
+      exit(-1);
```

Comment:
This means killing the user application, using an exception or just not catching it seems more user friendly to me.

Reply:
Done.